lib/ret/room_object.ex

```diff
@@ -0,0 +1,73 @@
+defmodule Ret.RoomObject do
+  use Ecto.Schema
+  import Ecto.Changeset
+  import Ecto.Query
+
+  alias Ret.{EncryptedField, Hub, RoomObject, Repo}
+  @schema_prefix "ret0"
+  @primary_key {:room_object_id, :id, autogenerate: true}
+
+  schema "room_objects" do
+    field(:room_object_sid, :string)
+    field(:gltf_node, EncryptedField)
+
+    belongs_to(:hub, Hub, references: :hub_id)
+
+    timestamps()
+  end
+
+  def perform_pin!(%Hub{hub_id: hub_id} = hub, %{room_object_sid: room_object_sid} = attrs) do
+    attrs = attrs |> Map.put(:gltf_node, attrs |> Map.get(:gltf_node) |> Poison.encode!())
+
+    room_object =
+      RoomObject
+      |> where([t], t.hub_id == ^hub_id and t.room_object_sid == ^room_object_sid)
+      |> preload(:hub)
+      |> Repo.one()
+
+    changeset(room_object || %RoomObject{}, hub, attrs) |> Repo.insert_or_update!()
+    Cachex.expire(:room_gltf, hub_id, -1)
+  end
+
+  def perform_unpin(%Hub{hub_id: hub_id}, room_object_sid) do
+    RoomObject
+    |> where([t], t.hub_id == ^hub_id and t.room_object_sid == ^room_object_sid)
+    |> Repo.delete_all()
+
+    Cachex.expire(:room_gltf, hub_id, -1)
+  end
+
+  def gltf_for_hub_id(hub_id) do
+    nodes =
+      RoomObject
+      |> where([t], t.hub_id == ^hub_id)
+      |> Repo.all()
+      |> Enum.map(&(&1.gltf_node |> Poison.decode!()))
+
+    node_indices =
+      if length(nodes) == 0 do
+        []
+      else
+        0..((nodes |> length) - 1) |> Enum.to_list()
+      end
+
+    gltf =
+      %{
+        asset: %{version: "2.0", generator: "reticulum"},
+        scenes: [%{nodes: node_indices, name: "Room Objects"}],
+        nodes: nodes,
+        extensionsUsed: ["HUBS_components"]
+      }
+      |> Poison.encode!()
+
+    {:commit, gltf}
+  end
+
+  defp changeset(%RoomObject{} = room_object, %Hub{} = hub, attrs) do
+    room_object
+    |> cast(attrs, [:room_object_sid, :gltf_node])
```

Comment:
This is odd. Why are we accepting an `sid` from the client?

Reply:
ah, i'm using "sid" in the sense I originally meant it (string identifier) not the way we've been using it up to this point (X character, server generated identifier.) I can rename this field to reduce the confusion -- this is a client-generated, string, unique key for the object. (Incidentally the network id generated by network-aframe)

Reply:
i don't think we need a server-generated string-based surrogate key for these objects given the use-case -- we're already relying/assuming clients are generating these keys so the server is just going to now have a role in "enforcing" this for pinned objects by disallowing duplicates in the db. i didn't implement any error handing paths to gracefully handle dupes, the channel will just die and reconnect if they send "pin" for a non-unique id. (not the websocket connection, the channel connection)

Reply:
the new name i'll suggest using is `object_id`

Reply:
Gotcha. `object_id` sounds okay. I'd suggest `client_object_id` maybe?

Reply:
I think I prefer `object_id` since once it's landed in our db, it's effectively a primary key for the object we may be passing around from that point forward to refer to it. it just incidentally is generated client side